File: examples/vercel-ai-research-agent/src/smoke.ts
/**
 * Smoke runner — proves the demo's tool path end-to-end against a running
 * SBO3L daemon WITHOUT an OpenAI key. Hardcodes the tool execute() calls
 * a reasonable agent would make so CI / Heidi can verify the integration
 * without burning LLM tokens.
 *
 * Usage:
 *   SBO3L_ALLOW_UNAUTHENTICATED=1 cargo run --bin sbo3l-server &
 *   npm install
 *   npm run smoke
 *
 * AC verification:
 *   - install verified (`npm install` resolves)
 *   - runs end-to-end in <30s
 *   - signed PolicyReceipt visible in console
 *   - audit_event_id printed
 */

import { dataFetchTool, buildSbo3lTool, defaultClient, KH_WORKFLOW_ID } from "./tools.js";
import { PolicyDenyError, SBO3LError } from "@sbo3l/vercel-ai";

// Fresh nonce + expiry per run. The daemon's protocol.nonce_replay guard
// rejects exact-duplicate (nonce, agent_id) tuples, so a static nonce
// only succeeds the first time the smoke runs against a given daemon.
// randomUUID is available on globalThis since Node 19; the fallback covers
// older Node 18 builds and exotic runtimes.
function freshNonce(): string {
  return globalThis.crypto?.randomUUID?.() ?? `nonce-${Date.now()}-${Math.random().toString(36).slice(2)}`;
}

const APRP = {
  agent_id: "research-agent-01",
  task_id: "demo-vercel-ai-smoke-1",
  intent: "purchase_api_call" as const,
  amount: { value: "0.05", currency: "USD" as const },
  token: "USDC",
  destination: {
    type: "x402_endpoint" as const,
    url: "https://api.example.com/v1/inference",
    method: "POST" as const,
  },
  payment_protocol: "x402" as const,
  chain: "base",
  provider_url: "https://api.example.com",
  expiry: new Date(Date.now() + 5 * 60 * 1000).toISOString(),
  nonce: freshNonce(),
  risk_class: "low" as const,
};

async function main(): Promise<void> {
  console.log(`▶ smoke: KH workflow target = ${KH_WORKFLOW_ID}\n`);

  // Step 1: data_fetch
  console.log("▶ tool: data_fetch (GitHub status — public, low-noise)");
  const fetchOut = await dataFetchTool.execute!(
    { url: "https://www.githubstatus.com/api/v2/status.json" },
    {},
  );
  if ("error" in fetchOut) {
    console.log(`  fetch warning: ${fetchOut.error} (continuing — SBO3L gate is the test)`);
  } else {
    console.log(`  ✓ HTTP ${fetchOut.status}`);
  }

  // Step 2: sbo3l_pay
  console.log("\n▶ tool: sbo3l_pay (APRP → SBO3L → KH adapter)");
  const client = defaultClient();
  const payTool = buildSbo3lTool(client);

  try {
    const receipt = await payTool.execute!(APRP, {});
    console.log(`✓ allow — execution_ref ${receipt.execution_ref ?? "(none)"}`);
    console.log(`  audit_event_id: ${receipt.audit_event_id}`);
    console.log(`  signature.algorithm: ${receipt.signature.algorithm}`);
    console.log(`  signature.key_id: ${receipt.signature.key_id}`);
    process.exit(0);
  } catch (err: unknown) {
    if (err instanceof PolicyDenyError) {
      console.log(`✗ ${err.decision} — ${err.denyCode}`);
      console.log(`  audit_event_id: ${err.auditEventId}`);
      process.exit(2);
    }
    if (err instanceof SBO3LError) {
      console.log(`✗ daemon error — ${err.code} (status ${err.status})`);
      process.exit(2);
    }
    throw err;
  }
}

main().catch((err: unknown) => {
  console.error(`error: ${err instanceof Error ? err.message : String(err)}`);
  process.exit(1);
});

File: examples/crewai-research-agent/sbo3l_crewai_demo/smoke.py
"""Smoke runner — proves the demo's tool path end-to-end against a running
SBO3L daemon WITHOUT crewai or an OpenAI key. Hardcodes the tool calls a
reasonable agent would make so CI / Heidi can verify the integration
without burning LLM tokens or pulling crewai's heavy dep tree.

Usage:
    SBO3L_ALLOW_UNAUTHENTICATED=1 cargo run --bin sbo3l-server &
    .venv/bin/pip install -e .
    .venv/bin/python -m sbo3l_crewai_demo.smoke

AC verification:
    - install verified (`pip install -e .` resolves the workspace deps)
    - runs end-to-end in <30s (no LLM round-trips, no crewai bootstrap)
    - signed PolicyReceipt visible in console
    - audit_event_id printed
"""

from __future__ import annotations

import json
import sys

from .tools import KH_WORKFLOW_ID, build_sbo3l_pay_func, default_client, fetch_url

APRP = {
    "agent_id": "research-agent-01",
    "task_id": "demo-crewai-smoke-1",
    "intent": "purchase_api_call",
    "amount": {"value": "0.05", "currency": "USD"},
    "token": "USDC",
    "destination": {
        "type": "x402_endpoint",
        "url": "https://api.example.com/v1/inference",
        "method": "POST",
    },
    "payment_protocol": "x402",
    "chain": "base",
    "provider_url": "https://api.example.com",
    "expiry": "2026-05-01T10:31:00Z",
    "nonce": "01HTAWX5K3R8YV9NQB7C6P2DGM",
    "risk_class": "low",
}


def main() -> int:
    print(f"▶ smoke: KH workflow target = {KH_WORKFLOW_ID}\n")

    print("▶ tool: data_fetch (GitHub status — public, low-noise)")
    fetch_out = json.loads(fetch_url("https://www.githubstatus.com/api/v2/status.json"))
    if "error" in fetch_out:
        print(f"  fetch warning: {fetch_out['error']} (continuing — SBO3L gate is the test)")
    else:
        print(f"  ✓ HTTP {fetch_out['status']}")

    print("\n▶ tool: sbo3l_payment_request (APRP → SBO3L → KH adapter)")
    with default_client() as client:
        sbo3l_pay = build_sbo3l_pay_func(client)
        decision_raw = sbo3l_pay(json.dumps(APRP))
    decision = json.loads(decision_raw)

    print("  envelope:")
    for k, v in decision.items():
        print(f"    {k}: {json.dumps(v)}")

    if decision.get("decision") == "allow":
        print(f"\n✓ allow — execution_ref {decision.get('execution_ref') or '(none)'}")
        print(f"  audit_event_id: {decision.get('audit_event_id') or '(unknown)'}")
        return 0

    if "error" in decision:
        print(f"\n✗ transport error — {decision['error']}")
        return 2

    print(
        f"\n✗ {decision.get('decision', '?')} — deny_code {decision.get('deny_code', '?')}"
    )
    return 2


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception as e:
        print(f"error: {e}", file=sys.stderr)
        sys.exit(1)
